{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import anndata\n",
    "import matplotlib.colors as clr\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import os\n",
    "import pandas as pd\n",
    "import random\n",
    "import scanpy as sc\n",
    "import sys\n",
    "from collections import defaultdict\n",
    "from sklearn.cluster import KMeans, MiniBatchKMeans\n",
    "\n",
    "import warnings\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "sc.settings.verbosity = 0\n",
    "\n",
    "random.seed(123)\n",
    "np.random.seed(123)\n",
    "os.environ[\"PYTHONHASHSEED\"] = \"123\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Colors\n",
    "color_dct = [\"#F56867\",\"#FEB915\",\"#C798EE\",\"#59BE86\",\"#7495D3\",\"#6D1A9C\",\"#15821E\",\"#3A84E6\",\"#997273\",\"#787878\",\"#DB4C6C\",\"#9E7A7A\",\"#554236\",\"#AF5F3C\",\"#93796C\",\"#F9BD3F\",\"#DAB370\",\"#877F6C\",\"#268785\"]\n",
    "color_cts = clr.LinearSegmentedColormap.from_list(\"magma\", [\"#000003\", \"#3B0F6F\", \"#8C2980\", \"#F66E5B\", \"#FD9F6C\", \"#FBFCBF\"], N=256)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/Users/chenyang/Library/CloudStorage/Dropbox/mcDETECT'"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Import model\n",
    "current_dir = os.getcwd()\n",
    "model_dir = os.path.abspath(os.path.join(current_dir, \"..\"))\n",
    "\n",
    "sys.path.insert(0, model_dir)\n",
    "from utils import *\n",
    "from model import mcDETECT, spot_granule\n",
    "sys.path.pop(0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# -------------------- Read data -------------------- #\n",
    "\n",
    "# Cells\n",
    "adata = sc.read_h5ad(\"data/adata.h5ad\")\n",
    "\n",
    "# Neurons\n",
    "adata_neuron = adata[adata.obs[\"cell_type\"].isin([\"Glutamatergic\", \"GABAergic\"])].copy()\n",
    "\n",
    "# Transcripts\n",
    "transcripts = pd.read_parquet(\"data/transcripts.parquet\")\n",
    "\n",
    "# Genes\n",
    "genes = pd.read_csv(\"data/genes.csv\")\n",
    "genes = list(genes.iloc[:, 0])\n",
    "\n",
    "# Negative control markers\n",
    "nc_genes = pd.read_csv(\"data/negative_controls.csv\")\n",
    "nc_genes = list(nc_genes[\"Gene\"])\n",
    "\n",
    "# Spots\n",
    "spots = sc.read_h5ad(\"data/spots.h5ad\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "20"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Markers\n",
    "syn_genes = [\"Camk2a\", \"Cplx2\", \"Slc17a7\", \"Ddn\", \"Syp\", \"Map1a\", \"Shank1\", \"Syn1\", \"Gria1\", \"Gria2\", \"Cyfip2\", \"Vamp2\", \"Bsn\", \"Slc32a1\", \"Nfasc\", \"Syt1\", \"Tubb3\", \"Nav1\", \"Shank3\", \"Mapt\"]\n",
    "len(syn_genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Initialize mcDETECT\n",
    "mc = mcDETECT(type = \"MERSCOPE\", transcripts = transcripts, syn_genes = syn_genes, nc_genes = nc_genes, eps = 1.5,\n",
    "              minspl = 3, grid_len = 1, cutoff_prob = 0.95, alpha = 10, low_bound = 3, size_thr = 4,\n",
    "              in_nucleus_thr = (0.1, 0.9), l = 1, rho = 0.1, s = 1, nc_top = 15, nc_thr = 0.1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sphere_x</th>\n",
       "      <th>sphere_y</th>\n",
       "      <th>sphere_z</th>\n",
       "      <th>layer_z</th>\n",
       "      <th>sphere_r</th>\n",
       "      <th>size</th>\n",
       "      <th>comp</th>\n",
       "      <th>in_nucleus</th>\n",
       "      <th>gene</th>\n",
       "      <th>brain_area</th>\n",
       "      <th>global_y_new</th>\n",
       "      <th>global_x_new</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>841.70602</td>\n",
       "      <td>2111.3860</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>1.099598</td>\n",
       "      <td>5.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0</td>\n",
       "      <td>Camk2a</td>\n",
       "      <td>FT</td>\n",
       "      <td>2225.470019</td>\n",
       "      <td>462.280283</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>839.20960</td>\n",
       "      <td>2117.4731</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>1.160026</td>\n",
       "      <td>7.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0</td>\n",
       "      <td>Camk2a</td>\n",
       "      <td>FT</td>\n",
       "      <td>2231.031143</td>\n",
       "      <td>458.764775</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>839.05195</td>\n",
       "      <td>2129.9065</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.684313</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0</td>\n",
       "      <td>Camk2a</td>\n",
       "      <td>FT</td>\n",
       "      <td>2243.248276</td>\n",
       "      <td>456.450483</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>842.63940</td>\n",
       "      <td>2129.9260</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>1.560930</td>\n",
       "      <td>7.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0</td>\n",
       "      <td>Camk2a</td>\n",
       "      <td>FT</td>\n",
       "      <td>2243.890434</td>\n",
       "      <td>459.980046</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>832.27235</td>\n",
       "      <td>2146.3137</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.829829</td>\n",
       "      <td>4.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0</td>\n",
       "      <td>Camk2a</td>\n",
       "      <td>FT</td>\n",
       "      <td>2258.228949</td>\n",
       "      <td>446.924800</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    sphere_x   sphere_y  sphere_z  layer_z  sphere_r  size  comp  in_nucleus  \\\n",
       "0  841.70602  2111.3860       0.0        0  1.099598   5.0   2.0           0   \n",
       "1  839.20960  2117.4731       0.0        0  1.160026   7.0   3.0           0   \n",
       "2  839.05195  2129.9065       0.0        0  0.684313   3.0   1.0           0   \n",
       "3  842.63940  2129.9260       0.0        0  1.560930   7.0   1.0           0   \n",
       "4  832.27235  2146.3137       0.0        0  0.829829   4.0   2.0           0   \n",
       "\n",
       "     gene brain_area  global_y_new  global_x_new  \n",
       "0  Camk2a         FT   2225.470019    462.280283  \n",
       "1  Camk2a         FT   2231.031143    458.764775  \n",
       "2  Camk2a         FT   2243.248276    456.450483  \n",
       "3  Camk2a         FT   2243.890434    459.980046  \n",
       "4  Camk2a         FT   2258.228949    446.924800  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Read granules\n",
    "granules = pd.read_csv(\"output/all_granules.csv\")\n",
    "granules.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Granule expression profile\n",
    "granule_adata = mc.profile(granules, genes = genes)\n",
    "granule_adata.write_h5ad(\"output/granule_adata_raw.h5ad\")\n",
    "granule_adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # t-SNE embedding\n",
    "# sc.pp.normalize_total(granule_adata, target_sum=1e4)\n",
    "# sc.pp.log1p(granule_adata)\n",
    "# sc.tl.pca(granule_adata, n_comps=10, svd_solver=\"auto\")\n",
    "# sc.tl.tsne(granule_adata, n_pcs=10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# granule_adata.write_h5ad(\"output/granule_adata_tsne.h5ad\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Granule expression profile\n",
    "granule_adata = sc.read_h5ad(\"output/granule_adata_tsne.h5ad\")\n",
    "\n",
    "cutoff = 6250\n",
    "granule_adata.obs[\"global_y_new\"] = cutoff - granule_adata.obs[\"global_y_new\"]\n",
    "\n",
    "granule_adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Use self-defined genes (should be the same as 1_gene_ranking_raw)\n",
    "genes_syn_pre = [\"Bsn\", \"Gap43\", \"Nrxn1\", \"Slc17a6\", \"Slc17a7\", \"Slc32a1\", \"Snap25\", \"Stx1a\", \"Syn1\", \"Syp\", \"Syt1\", \"Vamp2\", \"Cplx2\"]\n",
    "genes_syn_post = [\"Camk2a\", \"Dlg3\", \"Dlg4\", \"Gphn\", \"Gria1\", \"Gria2\", \"Homer1\", \"Homer2\", \"Nlgn1\", \"Nlgn2\", \"Nlgn3\", \"Shank1\", \"Shank3\"]\n",
    "genes_axon = [\"Ank3\", \"Nav1\", \"Sptnb4\", \"Nfasc\", \"Mapt\", \"Tubb3\"]\n",
    "genes_dendrite = [\"Actb\", \"Cyfip2\", \"Ddn\", \"Dlg4\", \"Map1a\", \"Map2\"]\n",
    "\n",
    "ref_genes = list(set(genes_syn_pre + genes_syn_post + genes_axon + genes_dendrite))\n",
    "ref_genes = list(set(ref_genes) & set(genes))\n",
    "len(ref_genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Subset data\n",
    "granule_adata_subset = granule_adata[:, ref_genes].copy()\n",
    "granule_adata_subset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# K-Means clustering\n",
    "data = granule_adata_subset.X.copy()\n",
    "if not isinstance(data, np.ndarray):\n",
    "    data = data.toarray()\n",
    "\n",
    "n_clusters = 15\n",
    "# kmeans = KMeans(n_clusters = n_clusters, random_state = 42, n_init = 25)\n",
    "kmeans = MiniBatchKMeans(n_clusters = n_clusters, batch_size = 5000, random_state = 123)\n",
    "kmeans.fit(data)\n",
    "granule_adata.obs[\"granule_subtype_kmeans\"] = kmeans.labels_.astype(str)\n",
    "\n",
    "desired_order = [f\"{i}\" for i in range(n_clusters)]\n",
    "granule_adata.obs[\"granule_subtype_kmeans\"] = pd.Categorical(granule_adata.obs[\"granule_subtype_kmeans\"], categories=desired_order, ordered=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Reference gene expression\n",
    "marker_genes = {\"pre-syn\": genes_syn_pre, \"post-syn\": genes_syn_post, \"dendrites\": genes_dendrite, \"axons\": genes_axon}\n",
    "\n",
    "expression_data = pd.DataFrame(granule_adata.X, columns=granule_adata.var_names, index=granule_adata.obs_names)\n",
    "kmeans_labels = granule_adata.obs[\"granule_subtype_kmeans\"]\n",
    "cluster_marker_avg = defaultdict(lambda: defaultdict(int))\n",
    "\n",
    "for cluster, genes in marker_genes.items():\n",
    "    num_genes = len(genes)\n",
    "    for gene in genes:\n",
    "        if gene in expression_data.columns:\n",
    "            gene_expression = expression_data[gene].groupby(kmeans_labels).sum()\n",
    "            for kmeans_cluster, expr_sum in gene_expression.items():\n",
    "                cluster_marker_avg[kmeans_cluster][cluster] += expr_sum / num_genes\n",
    "\n",
    "cluster_percentages = {}\n",
    "for kmeans_cluster, cluster_counts in cluster_marker_avg.items():\n",
    "    total_expression = sum(cluster_counts.values())\n",
    "    if total_expression > 0:\n",
    "        cluster_percentages[kmeans_cluster] = {cluster: (count / total_expression) for cluster, count in cluster_counts.items()}\n",
    "    else:\n",
    "        cluster_percentages[kmeans_cluster] = {cluster: 0 for cluster in cluster_counts.keys()}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cluster_labels = list(cluster_percentages.keys())\n",
    "marker_clusters = list(marker_genes.keys())\n",
    "\n",
    "plot_data = np.array([\n",
    "    [cluster_percentages[cluster].get(marker, 0) for marker in marker_clusters]\n",
    "    for cluster in cluster_labels\n",
    "])\n",
    "\n",
    "cluster_labels_int = [int(cl) for cl in cluster_labels]\n",
    "sorted_indices = np.argsort(cluster_labels_int)\n",
    "\n",
    "cluster_labels_sorted = [cluster_labels[i] for i in sorted_indices]\n",
    "plot_data_sorted = plot_data[sorted_indices]\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(8, 6))\n",
    "bottom = np.zeros(len(cluster_labels_sorted))\n",
    "\n",
    "for i, marker_cluster in enumerate(marker_clusters):\n",
    "    ax.bar(cluster_labels_sorted, plot_data_sorted[:, i], bottom=bottom, label=marker_cluster)\n",
    "    bottom += plot_data_sorted[:, i]\n",
    "\n",
    "ax.set_xlabel(\"mcDETECT Clusters\")\n",
    "ax.set_ylabel(\"Percentage of Marker Genes\")\n",
    "ax.grid(False)\n",
    "ax.legend(title=\"Type\", bbox_to_anchor=(1.25, 1), loc=\"upper right\")\n",
    "plt.tight_layout()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot granule subtype heatmap\n",
    "ref_genes_sorted = [\"Bsn\", \"Gap43\", \"Nrxn1\", \"Slc17a6\", \"Slc17a7\", \"Slc32a1\", \"Stx1a\", \"Syn1\", \"Syp\", \"Syt1\", \"Vamp2\", \"Cplx2\", \"Camk2a\", \"Dlg3\", \"Dlg4\", \"Gphn\", \"Gria1\", \"Gria2\", \"Homer1\", \"Homer2\", \"Nlgn1\", \"Nlgn2\", \"Nlgn3\", \"Shank1\", \"Shank3\", \"Cyfip2\", \"Ddn\", \"Dlg4\", \"Map1a\", \"Map2\", \"Ank3\", \"Nav1\", \"Nfasc\", \"Mapt\", \"Tubb3\"]\n",
    "granule_adata_subset.obs[\"granule_subtype_kmeans\"] = pd.Categorical(granule_adata.obs[\"granule_subtype_kmeans\"])\n",
    "granule_adata_subset.obs[\"granule_subtype_kmeans\"] = pd.Categorical(granule_adata_subset.obs[\"granule_subtype_kmeans\"], categories=cluster_labels_sorted, ordered=True)\n",
    "\n",
    "ax = sc.pl.heatmap(granule_adata_subset, var_names = ref_genes_sorted, groupby = \"granule_subtype_kmeans\", cmap = \"Reds\", standard_scale = \"var\", dendrogram = False, swap_axes = True, show = False, figsize = (10, 6))\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Manual subtyping\n",
    "pre_list = [\"4\", \"9\"]\n",
    "post_list = [\"0\", \"1\", \"3\"]\n",
    "den_list = []\n",
    "axon_list = []\n",
    "pre_post_list = [\"8\"]\n",
    "pre_den_list = [\"13\"]\n",
    "post_den_list = [\"6\", \"11\", \"12\", \"14\"]\n",
    "pre_post_den_list = [\"5\", \"7\", \"10\"]\n",
    "others_list = [\"2\"]\n",
    "\n",
    "subtype_dict = {\"pre-syn\": pre_list, \"post-syn\": post_list, \"pre & post\": pre_post_list, \"pre & den\": pre_den_list, \"post & den\": post_den_list, \"pre & post & den\": pre_post_den_list, \"dendrites\": den_list, \"axons\": axon_list, \"others\": others_list}\n",
    "granule_adata.obs[\"granule_subtype\"] = np.nan\n",
    "for i in subtype_dict.keys():\n",
    "    ind = pd.Series(granule_adata.obs[\"granule_subtype_kmeans\"]).isin(subtype_dict[i])\n",
    "    granule_adata.obs.loc[ind, \"granule_subtype\"] = i\n",
    "granule_adata.obs[\"granule_subtype\"] = pd.Categorical(granule_adata.obs[\"granule_subtype\"], categories=[\"pre-syn\", \"post-syn\", \"pre & post\", \"pre & den\", \"post & den\", \"pre & post & den\", \"dendrites\", \"axons\", \"others\"], ordered=True)\n",
    "granules[\"granule_subtype\"] = pd.Categorical(granule_adata.obs[\"granule_subtype\"], categories=[\"pre-syn\", \"post-syn\", \"pre & post\", \"pre & den\", \"post & den\", \"pre & post & den\", \"dendrites\", \"axons\", \"others\"], ordered=True)\n",
    "granule_adata.obs[\"granule_subtype\"] = granule_adata.obs[\"granule_subtype\"].cat.remove_unused_categories()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.DataFrame({\"Proportion\": (granule_adata.obs[\"granule_subtype\"].value_counts(normalize=True).sort_index() * 100).round(2)})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "granule_adata = assign_palette_to_adata(granule_adata, obs_key = \"granule_subtype\", cmap_name = \"tab10\")\n",
    "granule_adata = assign_palette_to_adata(granule_adata, obs_key = \"granule_subtype_kmeans\", cmap_name = \"tab20\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot granule subtypes t-SNE\n",
    "sc.set_figure_params(figsize = (8, 8))\n",
    "ax = sc.pl.tsne(granule_adata, color=\"granule_subtype_kmeans\", size=2, show=False)\n",
    "ax.grid(False)\n",
    "ax.set_xticks([])\n",
    "ax.set_yticks([])\n",
    "ax.set_xlabel(\"\")\n",
    "ax.set_ylabel(\"\")\n",
    "ax.set_title(\"\")\n",
    "frame_width = 1.5\n",
    "for spine in ax.spines.values():\n",
    "    spine.set_linewidth(frame_width)\n",
    "plt.savefig(\"output/granule_subtype_kmeans_tsne.jpeg\", dpi = 300, bbox_inches = \"tight\")\n",
    "plt.close()\n",
    "\n",
    "ax = sc.pl.tsne(granule_adata, color=\"granule_subtype\", size=2, show=False)\n",
    "ax.grid(False)\n",
    "ax.set_xticks([])\n",
    "ax.set_yticks([])\n",
    "ax.set_xlabel(\"\")\n",
    "ax.set_ylabel(\"\")\n",
    "ax.set_title(\"\")\n",
    "frame_width = 1.5\n",
    "for spine in ax.spines.values():\n",
    "    spine.set_linewidth(frame_width)\n",
    "plt.savefig(\"output/granule_subtype_tsne.jpeg\", dpi = 300, bbox_inches = \"tight\")\n",
    "plt.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot granule subtypes scatterplot\n",
    "sc.set_figure_params(figsize = (6, 9))\n",
    "ax = sc.pl.scatter(granule_adata, x=\"global_y_new\", y=\"global_x_new\", color=\"granule_subtype_kmeans\", size=1, show=False)\n",
    "ax.grid(False)\n",
    "ax.set_xticks([])\n",
    "ax.set_yticks([])\n",
    "ax.set_xlabel(\"\")\n",
    "ax.set_ylabel(\"\")\n",
    "ax.set_title(\"\")\n",
    "frame_width = 1.5\n",
    "for spine in ax.spines.values():\n",
    "    spine.set_linewidth(frame_width)\n",
    "plt.savefig(\"output/granule_subtype_kmeans_scatter.jpeg\", dpi = 300, bbox_inches = \"tight\")\n",
    "plt.close()\n",
    "\n",
    "sc.set_figure_params(figsize = (6, 9))\n",
    "ax = sc.pl.scatter(granule_adata, x=\"global_y_new\", y=\"global_x_new\", color=\"granule_subtype\", size=1, show=False)\n",
    "ax.grid(False)\n",
    "ax.set_xticks([])\n",
    "ax.set_yticks([])\n",
    "ax.set_xlabel(\"\")\n",
    "ax.set_ylabel(\"\")\n",
    "ax.set_title(\"\")\n",
    "frame_width = 1.5\n",
    "for spine in ax.spines.values():\n",
    "    spine.set_linewidth(frame_width)\n",
    "plt.savefig(\"output/granule_subtype_scatter.jpeg\", dpi = 300, bbox_inches = \"tight\")\n",
    "plt.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot granule subtype heatmap\n",
    "group_order = pre_list + post_list + pre_post_list + pre_den_list + post_den_list + pre_post_den_list + den_list + axon_list + others_list\n",
    "granule_adata_subset.obs[\"granule_subtype_kmeans\"] = pd.Categorical(granule_adata_subset.obs[\"granule_subtype_kmeans\"], categories=group_order, ordered=True)\n",
    "\n",
    "ax = sc.pl.heatmap(granule_adata_subset, var_names = ref_genes_sorted, groupby = \"granule_subtype_kmeans\", cmap = \"Reds\", standard_scale = \"var\", dendrogram = False, swap_axes = True, show = False, figsize = (10, 7))\n",
    "plt.savefig(\"output/granule_subtype_kmeans_heatmap.jpeg\", dpi = 300, bbox_inches = \"tight\")\n",
    "plt.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "granule_adata.write_h5ad(\"output/granule_adata.h5ad\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "granule_adata = sc.read_h5ad(\"output/granule_adata.h5ad\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "granules[\"granule_subtype\"] = pd.Categorical(granule_adata.obs[\"granule_subtype\"], categories=[\"pre-syn\", \"post-syn\", \"pre & post\", \"pre & den\", \"post & den\", \"pre & post & den\", \"dendrites\", \"axons\", \"others\"], ordered=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(613140, 13)"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Synapses\n",
    "syn_types = [\"pre-syn\", \"post-syn\", \"pre & post\", \"others\"]\n",
    "synapses = granules[granules[\"granule_subtype\"].isin(syn_types)]\n",
    "synapses.to_csv(\"output/synapses.csv\", index = 0)\n",
    "synapses.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot synapses\n",
    "syn_adata = granule_adata[granule_adata.obs[\"granule_subtype\"].isin(syn_types)].copy()\n",
    "syn_adata.obs[\"brain_area\"] = pd.Categorical(syn_adata.obs[\"brain_area\"], categories = [\"CTXsp\", \"FT\", \"HPF-CA\", \"HPF-DG\", \"HPF-SR\", \"Isocortex\", \"MB\", \"OLF\", \"TH\"], ordered = True)\n",
    "\n",
    "sc.set_figure_params(scanpy = True, figsize = (6, 9))\n",
    "ax = sc.pl.scatter(syn_adata, alpha = 1, x = \"global_y_new\", y = \"global_x_new\", color = \"brain_area\", palette = color_dct, size = 1, title = \" \", show = False)\n",
    "ax.grid(False)\n",
    "ax.grid(False)\n",
    "ax.set_xticks([])\n",
    "ax.set_yticks([])\n",
    "ax.set_xlabel(\"\")\n",
    "ax.set_ylabel(\"\")\n",
    "frame_width = 1.5\n",
    "for spine in ax.spines.values():\n",
    "    spine.set_linewidth(frame_width)\n",
    "plt.savefig(\"output/synapses.jpeg\", dpi = 300, bbox_inches = \"tight\")\n",
    "plt.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Subtype count in brain regions\n",
    "area_list = [\"Isocortex\", \"OLF\", \"HPF-CA\", \"HPF-DG\", \"HPF-SR\", \"CTXsp\", \"TH\", \"MB\"]\n",
    "synapse_subtypes = [\"pre-syn\", \"post-syn\"]\n",
    "\n",
    "filtered_df = synapses[synapses[\"brain_area\"].isin(area_list) & synapses[\"granule_subtype\"].isin(synapse_subtypes)]\n",
    "count_df = filtered_df.groupby([\"brain_area\", \"granule_subtype\"]).size().reset_index(name=\"count\")\n",
    "count_df = count_df[count_df[\"granule_subtype\"].isin(synapse_subtypes)]\n",
    "\n",
    "count_df.to_csv(\"output/synapse_subtype_by_brain_area.csv\", index = 0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 17667 × 290\n",
       "    obs: 'spot_id', 'global_x', 'global_y', 'global_y_new', 'global_x_new', 'region_labels', 'brain_area', 'indicator', 'gnl_count', 'gnl_radius', 'gnl_size', 'gnl_score'\n",
       "    var: 'genes'"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Spot-level synapse metadata\n",
    "spot_synapse_all = spot_granule(granule = synapses, spot = spots)\n",
    "spot_synapse_all"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Visualize spot-level synapse count\n",
    "spot_synapse_all.obs[\"log_gnl_count\"] = [np.log2(i + 1) for i in list(spot_synapse_all.obs[\"gnl_count\"])]\n",
    "\n",
    "sc.set_figure_params(scanpy = True, figsize = (6, 9))\n",
    "ax = sc.pl.scatter(spot_synapse_all[spot_synapse_all.obs[\"log_gnl_count\"] != 0].copy(), alpha = 1, x = \"global_y_new\", y = \"global_x_new\", color = \"log_gnl_count\", color_map = color_cts, title = \" \", size = 55, show = False)\n",
    "ax.grid(False)\n",
    "ax.set_xticks([])\n",
    "ax.set_yticks([])\n",
    "ax.set_xlabel(\"\")\n",
    "ax.set_ylabel(\"\")\n",
    "for spine in ax.spines.values():\n",
    "    spine.set_linewidth(False)\n",
    "plt.savefig(\"output/synapse_density.jpeg\", dpi = 300, bbox_inches = \"tight\")\n",
    "plt.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "synapses_pre = synapses[synapses[\"granule_subtype\"] == \"pre-syn\"].copy()\n",
    "synapses_post = synapses[synapses[\"granule_subtype\"] == \"post-syn\"].copy()\n",
    "synapses_pre.shape, synapses_post.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "spot_synapse_pre = spot_granule(granule = synapses_pre, spot = spots).copy()\n",
    "spot_synapse_post = spot_granule(granule = synapses_post, spot = spots).copy()\n",
    "spot_synapse_post.obs[\"global_y_new\"] += 6000\n",
    "spot_synapse_pre_post = anndata.concat([spot_synapse_pre, spot_synapse_post], axis = 0, merge = \"same\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "spot_synapse_pre_post.obs[\"log_gnl_count\"] = [np.log2(i + 1) for i in list(spot_synapse_pre_post.obs[\"gnl_count\"])]\n",
    "df = spot_synapse_pre_post.obs.copy()\n",
    "\n",
    "fig, ax = plt.subplots(figsize = (12, 9))\n",
    "ax.set_facecolor(\"black\")\n",
    "\n",
    "sc = ax.scatter(df[\"global_y_new\"], df[\"global_x_new\"], c = df[\"log_gnl_count\"], cmap = color_cts, s = 4, alpha = 1)\n",
    "ax.grid(False)\n",
    "ax.set_xticks([])\n",
    "ax.set_yticks([])\n",
    "ax.set_xlabel(\" \")\n",
    "ax.set_ylabel(\" \")\n",
    "for spine in ax.spines.values():\n",
    "    spine.set_linewidth(1.5)\n",
    "plt.savefig(\"output/synapse_density_pre_post.jpeg\", dpi=300, bbox_inches=\"tight\", facecolor=\"black\")\n",
    "plt.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Synapse density\n",
    "array1 = np.array(spot_synapse_all.obs[\"gnl_count\"])\n",
    "array2 = np.array(spot_synapse_all.obs[\"brain_area\"])\n",
    "density1 = []\n",
    "\n",
    "area_list = [\"Isocortex\", \"OLF\", \"HPF-CA\", \"HPF-DG\", \"HPF-SR\", \"CTXsp\", \"TH\", \"MB\", \"FT\"]\n",
    "for j in area_list:\n",
    "    temp = array1[array2 == j]\n",
    "    density1.append(np.sum(temp) / len(temp))\n",
    "density1_scaled = scale(density1)\n",
    "\n",
    "array3 = np.array(spot_synapse_all[:, spot_synapse_all.var[\"genes\"].isin(syn_genes)].X)\n",
    "array3 = array3.sum(axis = 1)\n",
    "array4 = np.array(spot_synapse_all.obs[\"brain_area\"])\n",
    "density2 = []\n",
    "\n",
    "for j in area_list:\n",
    "    temp = array3[array4 == j]\n",
    "    density2.append(np.sum(temp) / len(temp))\n",
    "density2_scaled = scale(density2)\n",
    "\n",
    "density4 = [1.71603565, 1.964351308, 2.052720791, 1.139278326, 99, 1.678527951, 1.082904337, 0.444031185, 0.0199885]\n",
    "density4_scaled = scale(density4)\n",
    "\n",
    "density_comparison = pd.DataFrame({\"area_list\": area_list, \"transcript\": density2, \"mcDETECT\": density1, \"ground_truth\": density4})\n",
    "density_comparison.to_csv(\"output/synapse_density_comparison.csv\", index = 0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Correlation with ground truth\n",
    "weights = []\n",
    "for i in area_list:\n",
    "    area = np.sum(spot_synapse_all.obs[\"brain_area\"] == i)\n",
    "    weights.append(area)\n",
    "\n",
    "density1_scaled = np.delete(density1_scaled, 4)\n",
    "density2_scaled = np.delete(density2_scaled, 4)\n",
    "density4_scaled = np.delete(density4_scaled, 4)\n",
    "weights.pop(4)\n",
    "\n",
    "print(weighted_corr(density1_scaled, density4_scaled, weights), weighted_corr(density2_scaled, density4_scaled, weights))\n",
    "print(weighted_spearmanr(density1_scaled, density4_scaled, weights), weighted_spearmanr(density2_scaled, density4_scaled, weights))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.4"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
